### PR TITLE
RFC 1: Keyring design doc

### LINKED ISSUE #382: Finish design spec for Keyring and KeyringService.
This spec is framing the Least Authority audit of the keyring code and the service. It should probably be in two parts, one in tallycash/hd-keyring (for the keyring itself) and one in this repository (for the service).

- [x] Initial writeup
- [ ] #434

### PR DESCRIPTION
**Pull request by Shadowfiend**
This ports the Coda doc that Least Authority is using for the keyring audit to an RFC structure with a small amount of additional context, and fills in an initial RFC directory structure + an RFC template.

Closes #382.

### CODE REVIEW

File: rfc/rfc-1-keyring-design.adoc [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+The proposed structure consists of two components, the `@tallycash/hd-keyring`
+package, which implements the core keyring abstraction, and the
+`KeyringService` in the `tally-background` package in the
+`@tallycash/tally-extension` repository, which mediates access to the core
+keyring abstraction for the extension.
+
+==== Core Keyring
+
+The hd-keyring package exports two primary constructs: the HDKeyring and the
+SerializedHDKeyring. The HDKeyring is designed to maintain the privacy of alle
```
Comment (karlmoll) [inline rfc/rfc-1-keyring-design.adoc:40]: alle ==> all ?

File: rfc/rfc-1-keyring-design.adoc [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+can no longer be restarted.
+
+Tally services communicate data outwards in two ways:
+
+* All services have a set of events they may broadcast. These are expected to
+  be viewable by any external entity, and should only carry public (to the rest
+  of the extension) data.
+* Service method calls may return data. This data is expected to only be
+  viewable by the caller, though generally any outsider is expected to be able
+  to call into the service, meaning the restriction on returned data is
```
Comment (karlmoll) [inline rfc/rfc-1-keyring-design.adoc:90]: This sentence is a little long ==> "into the service. This means the restriction"

File: rfc/rfc-1-keyring-design.adoc (hunk 1 above)
Comment (karlmoll) [inline rfc/rfc-1-keyring-design.adoc:40]: alle ==> all ?

File: rfc/rfc-1-keyring-design.adoc [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  effectively the same; namely, the caller should only receive sensitive data
+  they have proven they have access to.
+
+==== The `KeyringService`
+
+`KeyringService` provides access to one or more internal keyrings (``HDKeyring``s
+from the core keyring package). It also persists these keyrings when necessary
+and loads them from storage at unlock time. The keyring service can be locked
+or unlocked. When unlocked, it has direct access to `HDKeyring` instances and
+their data, including serialized mnemonics, and mediates access to those
```
Comment (karlmoll) [inline rfc/rfc-1-keyring-design.adoc:100]: Parenthesis around "including serialized mnemonics" might make this sentence a little bit more readable - I had to read it a couple times to realize what "and mediates access" was referring to

File: rfc/rfc-1-keyring-design.adoc [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+Web Crypto tools, and is designed to avoid hand-rolled encryption. Both the key
+derivation from the password and the encryption of serialized keyring data is
+performed using Web Crypto. Decryption is similarly managed by Web Crypto.
+
+As with `HDKeyring`, `KeyringService` protects access to cached key information
+and keyrings by using ECMAScript private variables so external observers cannot
+use JavaScript features to read the data.
+
+`KeyringService` allows generating keyrings of a fixed strength (256 bits) as
+well as importing legacy keyrings generated by Metamask (as 12-word seed
```
Comment (karlmoll) [inline rfc/rfc-1-keyring-design.adoc:115]: Metamask ==> MetaMask?

**Review by karlmoll — COMMENTED** [review]
Mostly LGTM - just added a couple typo/readability suggestions

**Comment by Shadowfiend** [comment]
@karlmoll I believe those are now fixed, have a look and merge if the rest feels good to you. @mhluongo may not have time to do a pass here, and he's already looked at the core of the content before in Coda.

**Review by mhluongo — DISMISSED** [review]
Looks even better in Markdown :)

File: rfc/rfc-1-keyring-design.adoc [hunk 5]
```diff
@@ -0,0 +1,144 @@
+:toc: macro
+
+= RFC 1: Keyring Design
+
+== Background
+
+The Tally wallet allows users to both view data associated with an account of
+theirs, and sign transactions on behalf of that account using private key
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply rfc/rfc-1-keyring-design.adoc:8]: ❓

**Review by mhluongo — APPROVED** [review]
Approval take 2

**Pull request merged**